File: game-service/src/main/java/io/github/xpakx/habitgame/battle/BattleService.java
package io.github.xpakx.habitgame.battle;

import io.github.xpakx.habitgame.battle.dto.BattleResponse;
import io.github.xpakx.habitgame.battle.dto.MoveRequest;
import io.github.xpakx.habitgame.battle.dto.MoveResponse;

import java.util.List;

public interface BattleService {
    BattleResponse start(Long expeditionId, Long userId);
    MoveResponse move(MoveRequest request, Long battleId, Long userId);
    MoveResponse prepare(MoveRequest request, Long battleId, Long userId);
    List<MoveResponse> endTurn(Long battleId, Long userId);
}

File: frontend/src/app/battle/battle.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { environment } from 'src/environments/environment';
import { ExpeditionSummary } from '../expedition/dto/expedition-summary';
import { JwtService } from '../common/jwt-service';
import { BattleResponse } from './dto/battle-response';
import { MoveResponse } from './dto/move-response';
import { MoveRequest } from './dto/mover-request';

@Injectable({
  providedIn: 'root'
})
export class BattleService extends JwtService {
  private apiServerUrl = environment.apiServerUrl;

  constructor(private http: HttpClient) { 
    super();
  }

  public getBattle(expeditionId: number): Observable<BattleResponse> {
    return this.http.get<BattleResponse>(`${this.apiServerUrl}/expedition/${expeditionId}/battle"`, { headers: this.getHeaders() });
  }

  public prepare(request: MoveRequest, battleId: number): Observable<MoveResponse> {
    return this.http.post<MoveResponse>(`${this.apiServerUrl}/battle/${battleId}/position"`, request, { headers: this.getHeaders() });
  }

  public move(request: MoveRequest, battleId: number): Observable<MoveResponse> {
    return this.http.post<MoveResponse>(`${this.apiServerUrl}/battle/${battleId}/move"`, request, { headers: this.getHeaders() });
  }

  public endTurn(battleId: number): Observable<MoveResponse[]> {
    return this.http.post<MoveResponse[]>(`${this.apiServerUrl}/battle/${battleId}/turn/end"`, { headers: this.getHeaders() });
  }
}
